tsl/src/continuous_aggs/refresh.c

```diff
@@ -0,0 +1,381 @@
+ *
+ * It is OK to refresh more than the given refresh window, but not less. Since
+ * we can only refresh along bucket boundaries, we need to adjust the refresh
+ * window to be inclusive in both ends to be able to refresh the given
+ * region. For example, if the dotted region below is the original window, the
+ * adjusted refresh window includes all four buckets shown.
+ *
+ * | ....|.....|..   |
+ */
+static InternalTimeRange
+compute_bucketed_refresh_window(const InternalTimeRange *refresh_window, int64 bucket_width)
+{
+	InternalTimeRange result = *refresh_window;
+	InternalTimeRange largest_bucketed_window =
+		get_largest_bucketed_window(refresh_window->type, bucket_width);
+
+	if (result.start <= largest_bucketed_window.start)
+		result.start = largest_bucketed_window.start;
+	else
+		result.start = ts_time_bucket_by_type(bucket_width, result.start, result.type);
+
+	if (result.end >= largest_bucketed_window.end)
+		result.end = largest_bucketed_window.end;
+	else
+	{
+		/* We get the time value for the start of the bucket, so need to add
+		 * bucket_width to get the end of it */
+		result.end = ts_time_bucket_by_type(bucket_width, result.end, result.type);
+		result.end = result.end + bucket_width;
+	}
+
+	return result;
+}
+
+/*
+ * Initialize the refresh state for a continuous aggregate.
+ *
+ * The state holds information for executing a refresh of a continuous aggregate.
+ */
+static void
+continuous_agg_refresh_init(CaggRefreshState *refresh, ContinuousAgg *cagg,
+							const InternalTimeRange *refresh_window)
+{
+	MemSet(refresh, 0, sizeof(*refresh));
+	refresh->cagg = *cagg;
+	refresh->cagg_ht = cagg_get_hypertable_or_fail(cagg->data.mat_hypertable_id);
+	refresh->refresh_window =
+		compute_bucketed_refresh_window(refresh_window, cagg->data.bucket_width);
+	refresh->partial_view.schema = &cagg->data.partial_view_schema;
+	refresh->partial_view.name = &cagg->data.partial_view_name;
+
+	/* Lock the continuous aggregate's materialized hypertable to protect
+	 * against concurrent refreshes. Only reads will be allowed. This is a
+	 * heavy lock that serializes all refreshes. We might want to consider
+	 * relaxing this in the future, e.g., we'd like to at least allow
+	 * concurrent refreshes that don't have overlapping refresh windows.
+	 *
+	 * Concurrent refreshes on the same continuous aggregate could be achieved
+	 * if we protect the aggregate with a UNIQUE constraint on the GROUP BY
+	 * columns. This would allow concurrent refreshes, but overlapping ones
+	 * might fail with, e.g., unique violation errors. Those could be
+	 * captured, however, and ignored when we know it means someone else just
+	 * did the same work.
+	 */
+	LockRelationOid(refresh->cagg_ht->main_table_relid, ExclusiveLock);
+}
+
+/*
+ * Finalize a refresh and cleanup state, e.g., release locks.
+ */
+static void
+continuous_agg_refresh_finish(const CaggRefreshState *refresh)
+{
+	UnlockRelationOid(refresh->cagg_ht->main_table_relid, ExclusiveLock);
```

Comment:
shouldn't this be released at the end of the transaction? Want to confirm that is the semantics here.

Reply:
Yes, good catch. Fixed.